### PR TITLE
Week 4 stephen omarrah

### PR DESCRIPTION
**Pull request by SOMarrah**
# Submitting Week 4 Assessment

### ✅ Completed Files
Mark the following as confirmation that the files in this repository are complete.
- [x] JavaScript Coding Challenges
- [x] Ruby Coding Challenges
- [x] Interview Questions

### 📝 Feedback
If you would like specific feedback on a particular problem please leave details here:
For my code-challenges I did not work on refactoring because I wanted to pseudocode for everything (Ruby), and refactored over there. I also included a research and tools section in most of my Pseudocode, which I think overall helped me finish the coding portion faster.

I think question 3 in Code-Challenges I could use some help with the spread operator. I wasn't sure how to access the full spread in the code block which could make refactored code simpler.

### CODE REVIEW

File: code-challenges.test.js [hunk 1]
```diff
@@ -15,35 +15,187 @@
...
+        //conditional checking current value vs previous value
+        if(newArr[i] !== newArr[i-1]){
+            //push current value if it is not the same as previous
+            noDupe.push(newArr[i])
+        }
+    }
+    //return the new array
+    return noDupe
+}
+//Hope I did the stretch challenge correctly! I was thinking something like this: (array, array2, ...array3).sort((b,a)=>b-a) would be possible but was unable to figure out how to do it.
```
Comment (WalkerTexas-coder) [inline code-challenges.test.js:193]: Yeah super close! The beauty of the spread operator is that it creates the ability to pass in arguments and we can use a tool to destructer their names if we want. 

a simplified version of this 
```javascript
const func = (...array) => {
    return array
}

func([1,2,3],[4,5,6],['hello'])

// output
(3) [Array(3), Array(3), Array(1)]
0: (3) [1, 2, 3]
1: (3) [4, 5, 6]
2: ['hello']
length: 3
```

File: code-challenges.test.js [hunk 2]
```diff
@@ -15,35 +15,187 @@
...
+  //takes an array as an input
+    //expected output is a subset of the array shuffled with the index 0 removed
+//use a tool delete/remove index 0 --.shift()
+  //then randomize the array
+    //shortened array
+        //make a random number to represent a random index in the array
+        //make the current value become = value at the random index
+        //make the value at the random index become = to the value at the current index
+            //the shuffling happens a number of times = to the length
+//tools and research-> filter is a good tool to return a subset of an array & is a built in method. Math.floor & Math.Random can be used to randomize the array a for loop could also iterate over the array. With a for loop I can  shift() always removes the first element, and then returns that element. .length can be used to find the number to be shuffled from the array
```
Comment (WalkerTexas-coder) [inline code-challenges.test.js:45]: Well done!

File: code-challenges.test.js [hunk 3]
```diff
@@ -15,35 +15,187 @@
...
+    let newArr = []
+    //sort the old array
+    array.sort((a,b) => a-b)
+    //push index 0 (the lowest number)
+    newArr.push(array[0])
+    //push the last index (the highest number)
+    newArr.push(array[array.length-1])
+    //return the new array of lowest and highest
+    return newArr
+}
```
Comment (WalkerTexas-coder) [inline code-challenges.test.js:102]: Let's go!

File: code-challenges.test.js [hunk 4]
```diff
@@ -15,35 +15,187 @@
...
+        // newArr[i] === newArr[i-1]?noDupe.push(newArr[i]):newArr possible ternary instead?
+        //conditional checking current value vs previous value
+        if(newArr[i] !== newArr[i-1]){
+            //push current value if it is not the same as previous
+            noDupe.push(newArr[i])
+        }
+    }
+    //return the new array
+    return noDupe
+}
```
Comment (WalkerTexas-coder) [inline code-challenges.test.js:192]: This is really good!

File: interview-questions.md [hunk 5]
```diff
@@ -7,52 +7,51 @@ Then, research the question to expand on your answer. Even if you feel you have
...
 
-  Researched answer:
+  Researched answer: A DOM event is something that is triggered by user interaction with the DOM (Document object model). The DOM is what the user sees and can interact with. In React, when that user does interact with the DOM, like clicking a button, the DOM is compared to the Virtual DOM and that specific component is updated. This is powerful because the whole DOM is not re-rendered or loaded just that specific component is. A specific example could be adding onClick to a button, when the button is clicked a DOM event is triggered and the logic is ran. Recently I have been using React-Three-Fiber to have 3d models render in the DOM. DOM events are literal interactions with 3d objects. The downside to this is that performance when loading is based on the user's hardware and could lead to limitations.
```
Comment (WalkerTexas-coder) [inline interview-questions.md:20]: Awesome work!

File: ruby-challenges.rb [hunk 6]
```diff
@@ -9,9 +9,58 @@
...
+    # else
+    #     "Please enter a valid number."
+    # end
+end
+
+p int_only2(num1) # output --> "7 is odd"
+p int_only2(num2) # output --> "42 is even"
+p int_only2(num3) # output --> "221 is odd"
+p int_only2('6') # output --> "Please enter a valid number."
+# further refactor could use something like number ~= /0 through 9/ to check strings containing integers
```
Comment (WalkerTexas-coder) [inline ruby-challenges.rb:57]: Ternary!

File: ruby-challenges.rb [hunk 7]
```diff
@@ -31,3 +105,50 @@
...
+p taco_cat2(palindrome_tester1) # output --> "Racecar is a tacocat...i mean palindrome!"
+p taco_cat2(palindrome_tester2) # output --> "Sorry, LEARN is not a palindrome..."
+p taco_cat2(palindrome_tester3) # output --> "Rotator is a tacocat...i mean palindrome!"
+p taco_cat2("taco Cat") #"taco Cat is a tacocat...i mean palindrome!"
+p taco_cat2("race car race car") #"race car race car is a tacocat...i mean palindrome!"
+p taco_cat2('r4c3c4r')# "r4c3c4r is a tacocat...i mean palindrome!"
+p taco_cat2("red rum, sir, is murder") #"red rum, sir, is murder is a tacocat...i mean palindrome!"
+p taco_cat2("Was it a cat I saw?")#"Was it a cat I saw? is a tacocat...i mean palindrome!"
+p taco_cat2("Sit on a potato pan, Otis.")#"Sit on a potato pan, Otis. is a tacocat...i mean palindrome!"
+p taco_cat2("sator arepo tenet opera rotas")#"sator arepo tenet opera rotas is a tacocat...i mean palindrome!"
```
Comment (WalkerTexas-coder) [inline ruby-challenges.rb:153]: Lol
This is more depth than I have ever seen on palindrome. Well done.

**Review by WalkerTexas-coder — APPROVED** [review]
Talk about a grand slam! You knocked it out of the park Stephen! Great job coming up with your own ideas for pseudo code. Your code also communicates your ideas effectively which is a great touch. Your interview answers are exceptional and the depth of your ruby code is more than I have ever seen to date. Amazing work. I want to see this level of depth again next week!

File: code-challenges.test.js (hunk 1 above)
Comment (SOMarrah) [reply code-challenges.test.js:193]: Oh thats awesome!

File: ruby-challenges.rb (hunk 7 above)
Comment (SOMarrah) [reply ruby-challenges.rb:153]: Yea i went on a deep dive there haha

**Pull request merged**